### PR TITLE
Publish 'artifacts.properties' using matrix params

### PR DESCRIPTION
**Pull request by jgsogo**
Changelog: Feature: Information in the `artifacts.properties` file is sent using matrix-params too when a package is uploaded to a server (if it has the capability). This will be the recommended way to send these properties to Artifactory (release TBD) to bypass Nginx blocking properties with periods.
Docs: https://github.com/conan-io/docs/pull/1487

On top of https://github.com/conan-io/conan/pull/6013

- [x] TODO: URL-encode matrix parameters
- [x] Conan-Server ignores all this key-value pairs
- [x] TODO: if we want to send a list of values for the same property, there are two alternatives:
   + Use `key=value1;value2` in the `artifacts.property` file.
   + ~~Modify the representation of the `artifacts.properties` in the sources to allow duplicated values ([see patch here](https://gist.github.com/jgsogo/8570ed51bea04ada7f6b622da607caf6))~~
- [x] TODO: do we want to add this capability to ConanServer?
- [x] conan-server doesn't support values with character `/` (Artifactory does). Maybe using a [special filter](https://bottlepy.org/docs/dev/routing.html#wildcard-filters) for the `matrix_params` part of the URL it will. :negative_squared_cross_mark: conan-server doesn't accept matrix_params (they are sent if the server returns a `matrix_params` capability)
- [x] Do we want an _opt-in_? Using `matrix_params` capability

#REVISIONS: 1
#TAGS: slow

### CODE REVIEW

File: conans/client/rest/client_routes.py [hunk 1]
```diff
@@ -1,30 +1,49 @@
...
 
 
 class ClientCommonRouter(object):
     """Search urls shared between v1 and v2"""
 
+    def __init__(self, base_url, artifacts_properties):
+        self.base_url = base_url
+
+        if artifacts_properties:
+            matrix_params = dict(_remove_put_prefix(artifacts_properties))
```
Comment (lasote) [inline conans/client/rest/client_routes.py]: What would happen if we send both?

Comment (jgsogo) [reply conans/client/rest/client_routes.py]: Artifactory adds all the `properties` in this order: (1) headers, (2) files and (3) matrix params. I think, but I need to check (I'll add it to a TODO list), that duplicated items are not stored/shown.

**Comment by jgsogo** [comment]
If we are going to use a _capability_ I would use a custom one to keep it easy for other servers in the wild: they might be already using `CHECKSUM_DEPLOY` for a different thing and they might not be prepared to receive matrix params. But that decision is out of my hands.

**Comment by lasote** [comment]
If they might not be prepared to receive matrix params your are breaking them all in this PR

**Comment by jgsogo** [comment]
That's the reason I suggested an opt-in, but one dedicated.

**Comment by lasote** [comment]
Mhhh. But I'm thinking the optin cannot be in the client, we need a new capability anyway, especially if we cannot send the properties by the old and the new mechanism. Given that the implementation in Artifactory has been necessary, introduce a "matrix_properties" capability makes sense. We can get rid of all the changes in the server, we won't break other servers and we can send only by one mechanism.

File: conans/test/functional/remote/put_properties_test.py [hunk 2]
```diff
@@ -1,45 +1,104 @@
...
         self.assertTrue(os.path.exists(props_file))
 
     def put_properties_test(self):
-        wanted_vars = {"MyHeader1": "MyHeaderValue1;MyHeaderValue2", "Other": "Value"}
+        test_server = TestServer()
+        servers = {"default": test_server}
+
+        wanted_vars = {"key0": "value",
+                       "key1": "with space",
+                       "key2": "with/slash",  # FIXME: Not supported in conan-server
```
Comment (lasote) [inline conans/test/functional/remote/put_properties_test.py]: Remove fixme?

File: conans/test/functional/remote/put_properties_test.py [hunk 3]
```diff
@@ -1,45 +1,104 @@
...
+        client.run("export . lasote/stable")
+        client.run("upload Hello0/0.1@lasote/stable -c")
+
+    def matrix_params_test(self):
+        test_server = TestServer(server_capabilities=[MATRIX_PARAMS, ])
+        servers = {"default": test_server}
+
+        wanted_vars = {"key0": "value",
+                       "key1": "with space",
+                       "key2": "with/slash",  # FIXME: Not supported in conan-server
```
Comment (lasote) [inline conans/test/functional/remote/put_properties_test.py]: remove fixme?

File: conans/client/rest/rest_client_v1.py [hunk 4]
```diff
@@ -117,12 +118,23 @@ def _get_file_to_url_dict(self, url, data=None):
         urls = self.get_json(url, data=data)
         return {filepath: complete_url(self.remote_url, url) for filepath, url in urls.items()}
 
+    def _add_matrix_params(self, urls):
+        if not self._matrix_params:
+            return urls
+
+        matrix_str = self.router._matrix_params_str
```
Comment (lasote) [inline conans/client/rest/rest_client_v1.py]: This "private" access to the router is a bit weird. Should this belong to the router? Not sure but something a bit weird.

**Pull request merged**